### ISSUE TITLE
Can't build Android AAB with AppCenter (Expo SDK 43)

### ISSUE THREAD

**Issue by TomCorvus** [expo/expo#14944]
### Summary

Hello 👋

On monorepo, I have updated Android files manually and updated Expo to 43, but I have an error when I build the Android application with AppCenter.

### Managed or bare workflow? If you have `ios/` or `android/` directories in your project, the answer is bare!

bare

### What platform(s) does this occur on?

Android

### Environment

```
  Expo CLI 4.12.10 environment info:
    System:
      OS: macOS 11.6
      Shell: 5.8 - /bin/zsh
    Binaries:
      Node: 14.17.6 - /usr/local/bin/node
      Yarn: 3.0.2 - /usr/local/bin/yarn
      npm: 8.1.1 - /usr/local/bin/npm
      Watchman: 2021.08.02.00 - /usr/local/bin/watchman
    Managers:
      CocoaPods: 1.11.2 - /usr/local/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 21.0.1, iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0
    IDEs:
      Android Studio: 2020.3 AI-203.7717.56.2031.7784292
      Xcode: 13.1/13A1030d - /usr/bin/xcodebuild
    npmPackages:
      expo: ^43.0.1 => 43.0.1 
      react: ^17.0.2 => 17.0.2 
      react-dom: 17.0.2 => 17.0.2 
      react-native: ~0.64.2 => 0.64.2 
      react-native-web: ^0.17.5 => 0.17.5 
    npmGlobalPackages:
      expo-cli: 4.12.10
    Expo Workflow: bare
```

### Reproducible demo or steps to reproduce from a blank project

To reproduce, build a monorepo and tried to build one of the Android application with AppCenter.
I can build successfully the AAB with Android Studio.

You can find below the error :
```
FAILURE: Build failed with an exception.

* Where:
Settings file '/Users/runner/work/1/s/packages/myapplication/android/settings.gradle' line: 3

* What went wrong:
A problem occurred evaluating settings 'MyApplication'.
> String index out of range: 0
```

And here the content fo settings.gradle 
```
rootProject.name = 'MyApplication'

apply from: new File(["node", "--print", "require.resolve('expo/package.json')"].execute().text.trim(), "../scripts/autolinking.gradle");
useExpoModules()

apply from: new File(["node", "--print", "require.resolve('@react-native-community/cli-platform-android/package.json')"].execute().text.trim(), "../native_modules.gradle");
applyNativeModulesSettingsGradle(settings)

include ':app'
```

**Comment by Kudo** [comment]
hi there! sorry for late reply. i just took some time to setup a monorepo on appcenter. i can repro the issue and have a fix. before the pr, can you help to verify if that works for you?

## packages/myapplication/android/settings.gradle
```diff
--- a/packages/myapplication/android/settings.gradle
+++ b/packages/myapplication/android/settings.gradle
@@ -1,9 +1,9 @@
 rootProject.name = 'myapplication'

-apply from: new File(["node", "--print", "require.resolve('expo/package.json')"].execute().text.trim(), "../scripts/autolinking.gradle");
+apply from: new File(["node", "--print", "require.resolve('expo/package.json')"].execute(null, rootDir).text.trim(), "../scripts/autolinking.gradle");
 useExpoModules()

-apply from: new File(["node", "--print", "require.resolve('@react-native-community/cli-platform-android/package.json')"].execute().text.trim(), "../native_modules.gradle");
+apply from: new File(["node", "--print", "require.resolve('@react-native-community/cli-platform-android/package.json')"].execute(null, rootDir).text.trim(), "../native_modules.gradle");
 applyNativeModulesSettingsGradle(settings)

 include ':app'
```

## expo modules
should use [patch-package](https://github.com/ds300/patch-package) to patch code inside node_modules.

### `packages/myapplication/patches/expo+43.0.1.patch`

```diff
diff --git a/packages/myapplication/patches/expo+43.0.1.patch b/packages/myapplication/patches/expo+43.0.1.patch
new file mode 100644
index 0000000..8e1b702
--- /dev/null
+++ b/packages/myapplication/patches/expo+43.0.1.patch
@@ -0,0 +1,8 @@
+diff --git a/node_modules/expo/scripts/autolinking.gradle b/node_modules/expo/scripts/autolinking.gradle
+index 184e1b5..0505395 100644
+--- a/node_modules/expo/scripts/autolinking.gradle
++++ b/node_modules/expo/scripts/autolinking.gradle
+@@ -1,2 +1,2 @@
+-def autolinkingPath = ["node", "--print", "require.resolve('expo-modules-autolinking/package.json')"].execute().text.trim()
++def autolinkingPath = ["node", "--print", "require.resolve('expo-modules-autolinking/package.json')"].execute(null, rootDir).text.trim()
+ apply from: new File(autolinkingPath, "../scripts/android/autolinking_implementation.gradle");
```

### `packages/myapplication/patches/expo-modules-autolinking+0.3.3.patch`

```diff
+diff --git a/node_modules/expo-modules-autolinking/scripts/android/autolinking_implementation.gradle b/node_modules/expo-modules-autolinking/scripts/android/autolinking_implementation.gradle
+index 6d9d202..ae5f09c 100644
+--- a/node_modules/expo-modules-autolinking/scripts/android/autolinking_implementation.gradle
++++ b/node_modules/expo-modules-autolinking/scripts/android/autolinking_implementation.gradle
+@@ -44,7 +44,7 @@ class ExpoAutolinkingManager {
+     String[] args = convertOptionsToCommandArgs('resolve', this.options)
+     args += ['--json']
+
+-    String output = exec(args)
++    String output = exec(args, projectDir)
+     Object json = new JsonSlurper().parseText(output)
+
+     cachedResolvingResults = json
+@@ -84,11 +84,11 @@ class ExpoAutolinkingManager {
+       args += '--empty'
+     }
+
+-    exec(args)
++    exec(args, project.rootDir)
+   }
+
+-  static String exec(String[] commandArgs) {
+-    Process proc = commandArgs.execute()
++  static String exec(String[] commandArgs, File dir) {
++    Process proc = commandArgs.execute(null, dir)
+     StringBuffer outputStream = new StringBuffer()
+     proc.waitForProcessOutput(outputStream, System.err)
+     return outputStream.toString()
```

**Comment by TomCorvus** [comment]
@Kudo Thanks for your reply!
For patches, I just have to copy/paste the code in .patch files? No need to remove or replace something?
And during the AppCenter build process, patches will be automatically apply with `"postinstall": "patch-package",` script?

I'm not familiar with patch-package.

**Comment by pilartomas** [comment]
@Kudo Thanks for the patch. I can confirm that it fixed my Appcenter build.

**Comment by TomCorvus** [comment]
Thanks @Kudo It's working for me too!

**Comment by Kudo** [comment]
we've published new packages including the fix from #15109. for a managed project or a new `expo init` bare project, the problem should be fixed. thanks for reporting this issue to us.
